meta_store.go

```diff
@@ -151,6 +161,168 @@ func (s *MetaStore) Delete(v *RequestVars) error {
 	return err
 }
 
+// AddLocks write locks to the store for the repo.
+func (s *MetaStore) AddLocks(auth, repo string, l ...Lock) error {
+	if !s.authenticate(auth) {
+		return newAuthError()
+	}
+
+	err := s.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(locksBucket)
+		if bucket == nil {
+			return errNoBucket
+		}
+
+		var locks []Lock
+		data := bucket.Get([]byte(repo))
+		if data != nil {
+			if err := json.Unmarshal(data, &locks); err != nil {
+				return err
+			}
+		}
+		locks = append(locks, l...)
+		sort.Sort(LocksByCreatedAt(locks))
+		data, err := json.Marshal(&locks)
+		if err != nil {
+			return err
+		}
+
+		return bucket.Put([]byte(repo), data)
+	})
+	return err
+}
+
+// Locks retrieves locks for the repo from the store
+func (s *MetaStore) Locks(auth, repo string) ([]Lock, error) {
+	if !s.authenticate(auth) {
+		return nil, newAuthError()
+	}
+
+	var locks []Lock
+	err := s.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(locksBucket)
+		if bucket == nil {
+			return errNoBucket
+		}
+
+		data := bucket.Get([]byte(repo))
+		if data != nil {
+			if err := json.Unmarshal(data, &locks); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return locks, err
+}
+
+// FilteredLocks return filtered locks for the repo
+func (s *MetaStore) FilteredLocks(auth, repo, path, cursor, limit string) ([]Lock, string, error) {
+	locks, err := s.Locks(auth, repo)
+	if err != nil {
+		return nil, "", err
+	}
+
+	if cursor != "" {
+		lastSeen := -1
+		for i, l := range locks {
+			if l.Id == cursor {
+				lastSeen = i
+				break
+			}
+		}
+
+		if lastSeen > -1 {
+			locks = locks[lastSeen:]
+		} else {
+			return nil, "", fmt.Errorf("cursor (%s) not found", cursor)
+		}
+	}
+
+	if path != "" {
+		var filtered []Lock
+		for _, l := range locks {
+			if l.Path == path {
+				filtered = append(filtered, l)
+			}
+		}
+
+		locks = filtered
+	}
+
+	if limit != "" {
+		size, err := strconv.Atoi(limit)
+		if err != nil {
+			return nil, "", errors.New("unable to parse limit amount")
+		}
+
+		size = int(math.Min(float64(len(locks)), 3))
+		if size < 0 {
+			return nil, "", nil
+		}
+
+		if size+1 < len(locks) {
+			return locks[:size], locks[size+1].Id, nil
+		}
```

Comment:
I think that this code does cover the case where `size` is `== len(locks)`, since it falls through into L269, but I might prefer something more obvious:

```go
var next string
if size + 1 < len(locks) {
        next = locks[size+1].Id
}

return locks[:size], next, nil
```

Which I think is more direct, and will only return a next cursor ID if that index is within the bounds of `locks`.